library/core/src/ptr/mut_ptr.rs

```diff
@@ -123,72 +123,64 @@ impl<T: ?Sized> *mut T {
     /// because all the bytes are dedicated to describing the address. Platforms which need to store
     /// additional information in the pointer may not support this operation, since the 'expose'
-    /// side-effect which is required for [`with_exposed_provenance_mut`][] to work is typically not
+    /// side-effect which is required for [`with_exposed_provenance_mut`] to work is typically not
     /// available.
     ///
-    /// It is unclear whether this method can be given a satisfying unambiguous specification. This
-    /// API and its claimed semantics are part of [Exposed Provenance][super#exposed-provenance].
+    /// This is an [Exposed Provenance][crate::ptr#exposed-provenance] API.
     ///
     /// [`with_exposed_provenance_mut`]: with_exposed_provenance_mut
     #[inline(always)]
-    #[unstable(feature = "exposed_provenance", issue = "95228")]
+    #[stable(feature = "exposed_provenance", since = "CURRENT_RUSTC_VERSION")]
     pub fn expose_provenance(self) -> usize {
-        // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
         self.cast::<()>() as usize
     }
 
-    /// Creates a new pointer with the given address.
+    /// Creates a new pointer with the given address and the [provenance][crate::ptr#provenance] of
+    /// `self`.
     ///
-    /// This performs the same operation as an `addr as ptr` cast, but copies
-    /// the *address-space* and *provenance* of `self` to the new pointer.
-    /// This allows us to dynamically preserve and propagate this important
-    /// information in a way that is otherwise impossible with a unary cast.
+    /// This is similar to a `addr as *mut T` cast, but copies
+    /// the *provenance* of `self` to the new pointer.
+    /// This avoids the inherent ambiguity of the unary cast.
     ///
     /// This is equivalent to using [`wrapping_offset`][pointer::wrapping_offset] to offset
     /// `self` to the given address, and therefore has all the same capabilities and restrictions.
     ///
-    /// This API and its claimed semantics are an extension to the Strict Provenance experiment,
-    /// see the [module documentation][crate::ptr] for details.
+    /// This is a [Strict Provenance][crate::ptr#strict-provenance] API.
     #[must_use]
     #[inline]
-    #[unstable(feature = "strict_provenance", issue = "95228")]
+    #[stable(feature = "strict_provenance", since = "CURRENT_RUSTC_VERSION")]
     pub fn with_addr(self, addr: usize) -> Self {
-        // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
-        //
-        // In the mean-time, this operation is defined to be "as if" it was
-        // a wrapping_offset, so we can emulate it as such. This should properly
-        // restore pointer provenance even under today's compiler.
+        // This should probably be an intrinsic to make it more efficient, but meanwhile, we can
```

Comment:
I'm being surprised by the degree to which people are analyzing these comments in other venues, so I'm commenting on them here.

In what sense could this be more efficient? If you mean compile time, it would be nice to indicate that specifically. If you are referring to the fact that this needs to call `self.addr()` and therefore currently lowers to a `ptrtoint`, I'd prefer that be commented-on in the `addr` implementation or in cg_ssa instead, because the efficiency problem is in LLVM specifically.

Reply:
Another thing worth commenting on is the relative prevalence of this function specifically. So, if this is slow to compile, but is only present a couple of times in a code base, that's not as big of a deal.

It's definitely going to be hard to gauge these things until it's stable and more widely used, though.

Reply:
I am referring to the fact that this does pointer arithmetic, which is completely unnecessary. LLVM will likely optimize the arithmetic away, but it's still silly.

Reply:
> If you are referring to the fact that this needs to call self.addr() and therefore currently lowers to a ptrtoint, I'd prefer that be commented-on in the addr implementation or in cg_ssa instead, because the efficiency problem is in LLVM specifically.

That could be a comment in codegen, not in lib -- lib is doing everything right here by using `transmute` instead of `as`.